src/Interpreters/InterpreterSelectQuery.cpp

```diff
@@ -1723,6 +1749,67 @@ void InterpreterSelectQuery::executeAggregation(QueryPipeline & pipeline, const
     /// Forget about current totals and extremes. They will be calculated again after aggregation if needed.
     pipeline.dropTotalsAndExtremes();
 
+    /// TODO better case determination
+    if (group_by_info && settings.optimize_aggregation_in_order)
+    {
+        auto & query = getSelectQuery();
+        SortDescription group_by_descr = getSortDescriptionFromGroupBy(query, *context);
+        bool need_finish_sorting = (group_by_info->order_key_prefix_descr.size() < group_by_descr.size());
+
+        if (need_finish_sorting)
+        {
+            /// TOO SLOW
+        }
+        else
+        {
+            if (pipeline.getNumStreams() > 1)
+            {
+                auto many_data = std::make_shared<ManyAggregatedData>(pipeline.getNumStreams());
+                size_t counter = 0;
+                pipeline.addSimpleTransform([&](const Block & header)
+                {
+                    return std::make_shared<AggregatingInOrderTransform>(header, transform_params, group_by_descr, settings.max_block_size, many_data, counter++);
+                });
+
+                /// TODO remove code duplication
+                for (auto & column_description : group_by_descr)
```

Comment:
But why we need to do something special with `group_by_desc` here? Exactly the same code will be run at creation of `AggregatingSortedTransform`.

Reply:
Actually it won't https://github.com/ClickHouse/ClickHouse/blob/1a4ce6b3c806f77e548dacf4207e06542231fa19/src/Processors/Merges/Algorithms/IMergingAlgorithmWithDelayedChunk.cpp#L13
and will fail here
https://github.com/ClickHouse/ClickHouse/blob/1a4ce6b3c806f77e548dacf4207e06542231fa19/src/Core/SortCursor.h#L66

Reply:
You are right. I confused `MergingSortedAlgorithm` with `MergingSortedAlgorithm`.